### DISCUSSION TITLE
BRouter external profiles

### DISCUSSION THREAD

**Discussion by Leeb64** [devemux86/cruiser#266] [Q&A] [answered]
Would it be possible to save the BRouter profile in the preferences?
Among other things, the external router profile used.
This allows you to save different presets for different vehicle types.
You would save yourself the whole reconfiguration and simply call up a different preset.
The external profile is currently not saved.
The internal profile always comes up when you call up the saved profile.

**Comment by devemux86** [comment, marked as answer]
Any file access on new Androids requires first the user permission, via Android file browser.
As it happens when we import / export, load BRouter profile / no-go area, external theme, ...

I might look into it more, but new Androids are making the lives of users (and developers)
more and more difficult...

**Reply by 0709wiwiwi** [reply]
@Leeb64 
Use an external Brouter profile and then switch back is too complex....this deters you from taking advantage of this. 
You also thought of a Preset solution by multiple .pref files ?  https://github.com/devemux86/cruiser/discussions/269
Just had a test and bad luck, the Cruiser Brouter settings are indeed not restored with it.
_I doubt but wasn't this possible in a single previous version with the  .brf files were placed in a Cruiser folder in Android media ?_

**Reply by devemux86** [reply]
@0709wiwiwi the external files (profiles, maps, themes) can be anywhere on the storage.
And they require user permission that is granted by the Android file browser. Google rules.

The whole situation is quite complicated and it involves power using.
So I may check it when there is available time from all other features. :slightly_smiling_face:

**Reply by 0709wiwiwi** [reply]
Emux, I don't need to use this so I just briefly tested the external profile option.  
Turning this back to the Cruiser "default" after that is already confusing enough.

**Reply by 0709wiwiwi** [reply]
That I want to avoid to use external Brouter profiles. It's too confusing, as once you use an external Brouter profile, the routing selection switchover Car > Motor > etc...doesn't work either which is quite surprising.  Anyway I just like to ask for a single extra simple Cruiser Brouter profile.  I will open a new idea request later.

**Reply by 0709wiwiwi** [reply]
Listen to other opinions first please. Maybe other people can handle it all fine.  I just do share you my experience.  

I do not have trouble with open external maps nor external themes nor Brouter no-go areas. 
 Why external profiles are probably still usefull and are appreciated by some users ?  
Example. 
The Cruiser_mtb profile is a trekking bike profile, not as fine tuned as the Zossebart mtb profiles.  (wet - hard etc)

**Reply by Leeb64** [reply]
I made a motorcycle profile out of a Poutnik profile...
In contrast to the internal profile, it reliably bypasses towns and cities if, for example, there is a bypass road.
The rest of the route is pretty similar with the internal profile optimally set.
This bypass can be reached by changing the "highway=crossings" parameter.
it is not available in the driving cruiser profile..
I'll attach my profile here.
The highway=crossing parameter is at the bottom at the end.

[___MOPPED xxx.txt](https://github.com/devemux86/cruiser/files/12258871/___MOPPED.xxx.txt)
(change "txt" to "brf")

**Reply by devemux86** [reply]
@Leeb64 thanks for the profile.

We used them in the past, but after the BRouter fix and recommendation by @poutnikl we returned to BRouter profiles:
- https://github.com/abrensch/brouter/issues/568#issuecomment-1586242723
- https://github.com/devemux86/cruiser/discussions/175#discussioncomment-6179216

**Reply by devemux86** [reply]
I do not see any real difference or do not understand how it is supposed to work.

Same or more detailed routing exists with driving-cruiser profile and use avoids for motorways, primary roads, ..., etc.

**Reply by Leeb64** [reply]
Here is an example int vs ext

<img src="https://github.com/devemux86/cruiser/assets/120722053/cf074920-237b-4072-a4e4-528b63e980af" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/120722053/0a44bcb7-df39-413a-907b-c4f855fca987" width="50%"/>

**Reply by Leeb64** [reply]
Here, as in many other cases, unnecessary annoying through-town driving is avoided.

Avoidance:
Tolls, ferries, residential roads, dirt roads, avoid cities, 70km/h, 100kg

**Reply by devemux86** [reply]
What options do you use in internal?

**Reply by devemux86** [reply]
Anyway, the image shows that the external seems to pass through residential areas.

**Reply by DolomitesTom** [reply]
I need to agree, I have activated avoid cities and avoid residential streets and not excluded main roads. I would also expect it to use the main road, D1059 in this case.
The internal Profile uses residential roads.

**Reply by devemux86** [reply]
BRouter routing can be easily tested in BRouter website, that provides "Analyze route" and "Data table" tools.

This is the only way to verify if a calculated route contains a road type or not. :slightly_smiling_face: 

> The internal Profile uses residential roads.

Please provide a route and the options and we can verify the tags in BRouter website.

**Reply by DolomitesTom** [reply]
To my point of view the avoid cities option should also be valid for small villages of 1000 habitants. 

Or we need something like prefer detours of up to X kilometers.

Interesting is that the option avoid Traffic uses a long detour.

**Reply by DolomitesTom** [reply]
What is this symbol to upload a file on the mobile phone?

**Reply by DolomitesTom** [reply]
48.31657640578043,7.25944457303314
https://plus.codes/8FW98785+JQ

Destination https://plus.codes/8FW955HX+X5
48.17993100000001,7.197901999999998

**Reply by devemux86** [reply]
> To my point of view the avoid cities option should also be valid for small villages of 1000 habitants.
> Or we need something like prefer detours of up to X kilometers.

Everyone, better express any routing reports in [BRouter repository](https://github.com/abrensch/brouter/) or they will stay here unread! :slightly_smiling_face:

**Reply by devemux86** [reply]
Everyone, if you expect someone to check any BRouter reports, use the [BRouter website](https://brouter.de/brouter-web/).
Load driving-cruiser profile, calculate a route and post the link and the activated options.

Then we can examine the result using BRouter website tools.

**Reply by DolomitesTom** [reply]
How do I load the internal driving-cruiser profile in Brouter website or how can I use the new option like avoid cities?

**Reply by devemux86** [reply]
> How do I load the internal driving-cruiser profile in Brouter website or how can I use the new option like avoid cities?

See the instructions in:

- https://github.com/devemux86/cruiser/discussions/272

**Comment by DolomitesTom** [comment]
ok have made a post in the other repository https://github.com/abrensch/brouter/issues/608#issue-1836846261

**Reply by devemux86** [reply]
> ok have made a post in the other repository

Better to provide them with BRouter links.

Developers should not be expected to have free time to reproduce reports from images or coordinates.
They do not know everyone's area. They need the BRouter links to see the report instantly with 1 click.

**Comment by DolomitesTom** [comment]
Kuviger by default also calculates the D459 instead of D1059 with the option curvy roads (independent of any avoidance strategies), but with the option fast and curvy road it uses the D1059. Brouter website uses the D1059 in the 1st alternative. But what are the options for the Brouter 1st alternative? Would be interesting to know if something could be done in options to force the 1st alternative as original route.

**Reply by devemux86** [reply]
Or skip the "Avoid cities" option and just use the many "Avoid X roads" options:
motorways, primary roads, ...

**Reply by DolomitesTom** [reply]
in this case it would not help since it should use the main road D1059.

**Reply by devemux86** [reply]
The road has some tags, check them on BRouter or OpenStreetMap websites
and use the specific avoid option to avoid this road.

**Comment by DolomitesTom** [comment]
When I calculate the same route with Kuriger 1 and use the fastest routing option, then the D1059 is used. In the new routing options even fast car profile don't use it. How can the fastest routing be forced with motorcycle?

**Reply by devemux86** [reply]
Reset the BRouter options to default and _do not change the speed_ (in Advanced).
(default speed is 160 as seen in BRouter website)

This is the normal motor routing. Then you can play with the various avoid options.

**Reply by DolomitesTom** [reply]
Did you see the answer in Brouter repository? I do not understand it because I don't know what each town class represents, but the result seems to work. Could you "translate" it in an option avoid small villages in the internal cruiser profile?

**Reply by devemux86** [reply]
They mention that BRouter currently only considers towns with population > 50K, see:
- https://github.com/abrensch/brouter/issues/486#issuecomment-1646872290

Regarding their [suggestion](https://github.com/abrensch/brouter/issues/608#issuecomment-1666080706) with the `estimated_town_class` I could check it.
But it seems to require an updated E5_N45.rd5 so I cannot test it easily on BRouter website.

Perhaps we should wait, they discuss for more improvements.

**Reply by devemux86** [reply]
@quaelnix suggested a nice profile change, I will check it:

- https://github.com/abrensch/brouter/issues/608#issuecomment-1666424408

**Comment by devemux86** [comment]
I updated the `driving-cruiser` profile with the above suggestion, at least the reported [route](https://brouter.de/brouter-web/#map=14/48.2593/7.2437/standard&lonlats=7.274363,48.277729;7.209955,48.245768&profile=car-fast) now works.

Please test if the "Avoid cities" works better. Instructions are here:
- https://github.com/devemux86/cruiser/discussions/272

**Comment by DolomitesTom** [comment]
I don't find the option "avoid cities". Is it what is called "consider town" in Brouter?

If so, then yes this single example avoids now the little villages and uses the main road instead. But I don't know if this has a positive effect on the avoid cities in general.

When I for example change the starting point, then it drives a boring route instead of using the pretty Cols in the Voges.

Within all these avoid option I am missing the simple "use fastest" or use "shortest route".

![grafik](https://github.com/devemux86/cruiser/assets/121371629/fa1af42e-57a9-4585-85fd-a7e590b23cbd)

I thought these new avoid rules would lead to curvy routing. This consider forests is not working at all. The Voges is in principle one big forest. But using this it drives towards Strasbourg and uses absolute boring streets on the other side instead of driving the Col's.

![grafik](https://github.com/devemux86/cruiser/assets/121371629/f65f4831-6c90-4cc3-8f75-9ddb88d95a62)

**Reply by devemux86** [reply]
Everyone please let's try to follow some simple rules when we make reports here or in BRouter: https://github.com/devemux86/cruiser/discussions/266#discussioncomment-6639735

And please do not copy the whole profile text, attach the file.

> I don't find the option "avoid cities". Is it what is called "consider town" in Brouter?

:+1:

**Comment by DolomitesTom** [comment]
> Everyone please let's try to follow some simple rules when we make reports here or in BRouter

sorry, don't understand what you mean, which simple rule?

Calculating the same (no route, just any starting point near Saverne and any destination near Munster in the Voges) in Kurviger with option "fast and curvy" makes all what is needed. It uses the Col's and also the D1059. I don't find any setting in Brouter that leads to a similar route calculation. Either it takes huge detours around to the left on roads I would never take or around to the right, equally boring routes.

Which settings are the right ones with the latest internal profile to just use the "quickest" or "shortest" way without motorways?

![grafik](https://github.com/devemux86/cruiser/assets/121371629/a6e88a71-1e45-4339-a6e1-08ce3b121930)

**Reply by devemux86** [reply]
> sorry, don't understand what you mean, which simple rule?

> Developers should not be expected to have free time to reproduce reports from images or coordinates.
They do not know everyone's area. They need the BRouter links to see the report instantly with 1 click.

---

Result and as mentioned before:
Let's use the familiar avoidance options for road types that some websites use and advertise with fancy names.

**Reply by DolomitesTom** [reply]
ok understand, just the URL from BRouter is also sufficient?

**Reply by DolomitesTom** [reply]
Does this help?

https://brouter.de/brouter-web/#map=9/48.4393/6.9512/standard&lonlats=7.2472,48.74624;7.225571,48.22879  
Do I need to copy back the brouter profile from BRouter Website into a file and then upload here? or are the settings per screenshot ok?

![grafik](https://github.com/devemux86/cruiser/assets/121371629/333761d5-b291-41b0-a17e-95fa415d6d55)
![grafik](https://github.com/devemux86/cruiser/assets/121371629/5e80c36e-7199-4c8b-aaa9-dfc7f45cf589)

**Reply by devemux86** [reply]
- BRouter website route link
- What profile was used
Attach file (not copy text) if made changes
- Write the activated options

**Reply by DolomitesTom** [reply]
consider town
avoid motorways
avoid urban
ignore barrier
valid for motorcycle
vmax = 70
[driving-cruiser from Brouter website test.zip](https://github.com/devemux86/cruiser/files/12267430/driving-cruiser.from.Brouter.website.test.zip)

this is what BRouter makes
![grafik](https://github.com/devemux86/cruiser/assets/121371629/e6f37a09-1cca-45fe-8f6b-9548b49bc24c)

here the kurviger file
[Ueber den Felsen, Pfalzburg - Markirch.zip](https://github.com/devemux86/cruiser/files/12267438/Ueber.den.Felsen.Pfalzburg.-.Markirch.zip)

**Reply by devemux86** [reply]
Quick note: as mentioned before, do not change the default options, use default vmax=160.

**Reply by devemux86** [reply]
> Does this help?
https://brouter.de/brouter-web/#map=9/48.4393/6.9512/standard&lonlats=7.2472,48.74624;7.225571,48.22879

:+1: 

> Do I need to copy back the brouter profile from BRouter Website into a file and then upload here?

If you do not make internal changes in the profile code, just mention the profile name.
(if you made changes, attach the profile file)

> or are the settings per screenshot ok?

Or more simple, just write the activated options (as you did above).

**Reply by devemux86** [reply]
> Within all these avoid option I am missing the simple "use fastest" or use "shortest route".

The fastest route is with the _default options_: all avoid options are off -> use all / fastest roads.

And do not reduce vmax from default 160.

**Comment by quaelnix** [comment]
@devemux86, In my honest opinion, both the new forest and city tags are not ready for production, but if you want to use them anyway, I would suggest changing:
```
switch estimated_forest_class= 0 99 0
```
into something like:
```
switch not estimated_forest_class=4|5|6 99 0 0
```
and:
```
min if greater no_forest_penalty 0 then 1 else 999
```
into something like:
```
min if greater no_forest_penalty 0 then 25 else 999
```
The "mean cost factor" should be kept as close to 1 as possible.

**Reply by devemux86** [reply]
@quaelnix thanks for the help!

I will check it and also move these options into the "Advanced" section in Cruiser settings.

These new BRouter options are quite popular in users, but we need to understand how to use them.
"Avoid road X" options are more understandable and we can verify them with BRouter website tools.

**Comment by DolomitesTom** [comment]
IThe "problem" or let's say difference of Brouter is that the avoidance of primary, secondary or teritiary roads is always interpreted as almost strict avoidance. When you for example just have activated to avoid primary roads, then you find some nice secondary and other roads, but if you then want to use a primary road for a certain section, it stubbornly refuses to use them and then leads to chaotic detours if you set some shaping points on it.

In order to get some smoth results, one would need something like mitigated levels of avoidance from 0 to 6 e.g. from no avoidance over weak avoidance till strong avoidance for each of these road types, maybe also for these new fancy options (if they work).

**Reply by devemux86** [reply]
They are not strict as in default BRouter profiles.
And they're not entirely lenient avoidances either.

---

That's why `driving-cruiser` profile is public so that everyone can study / test it and suggest improvements.
Based on your experience and tests on BRouter website we can certainly make it better. :slightly_smiling_face: 

A new version is available, with the forest improvement suggested above by @quaelnix:
- https://github.com/devemux86/cruiser/discussions/272

**Reply by DolomitesTom** [reply]
Yes I had used this

**Reply by devemux86** [reply]
> Yes I had used this

The profile? You could not, as it was just updated:

> A new version is available, with the forest improvement suggested above by @quaelnix

**Reply by DolomitesTom** [reply]
just tested again, hardly any difference.

**Reply by devemux86** [reply]
There are many differences in calculated routes, that's why the change was made.

When we use the forests option, where the change was made. :slightly_smiling_face:

**Reply by quaelnix** [reply]
@DolomitesTom, there is a major difference, compared to how it used to be:
![consider_forest](https://github.com/devemux86/cruiser/assets/122357328/001b189b-9428-497e-ae57-f254ad6a9888)

**Reply by DolomitesTom** [reply]
yes, this looks much better. Thanks, have to try again.

**Reply by devemux86** [reply]
We are all learning how to make BRouter profiles better and better.
It's amazing how much power / flexibility their profile language has.

We just need to invest some time for tests and try different things. :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
> They are not strict as in default BRouter profiles.
And they're not entirely lenient avoidances either.

Here within Cruiser
consider town
avoid motorways
avoid urban
valid for motorcycle
avoid primary
avoid upaved

vmax = 70
[Test Voges avoid primary.zip](https://github.com/devemux86/cruiser/files/12267529/Test.Voges.avoid.primary.zip)

Example here: I wanted to set 2 shaping points to use the primary D604 to drive to direction Saverne (just for example, there are some nice curves on the primary road)

Instead of using this D604 it calculates a chaotic detour. This looks to me as almost strict avoidance. Ok I can set tons of additional shaping points maybe to force the D604, but this in not intuitive.

I tried the same on the BRouter Website. Yes, it is not a 100% strict avoidance, but a very strong avoidance. You need serveral shaping points to force to use the primary road if these are avoided.
This is what I mean. If there was a setting with weak avoidance of primary road, then you maybe need 1 or 2 to force the usage of a primary road in a certain section if you want to.

![grafik](https://github.com/devemux86/cruiser/assets/121371629/be86e863-9aee-4216-81e1-1b3e50ed8dfd)

**Reply by devemux86** [reply]
> vmax = 70

You are doing the same mistake again and again, please follow the above suggestions and only change the needed options.

> And do not reduce vmax from default 160.

**Reply by devemux86** [reply]
> vmax = 70

This "tries" to avoid roads with speed > 70.

**Reply by DolomitesTom** [reply]
> You are doing the same mistake again and again, please follow the above suggestions and only change the needed options.

obviously 🤦‍♂️ stubborn Jack 🙈.... 😄

> This "tries" to avoid roads with speed > 70.

good to know, then I have obviously misunderstood 😱, brain barrier just broken 🤣... sorry for the inconvenience

**Reply by devemux86** [reply]
`vmax` is a clever (perhaps more advanced) way to avoid high speed roads instead of using the avoidance options.

**Comment by quaelnix** [comment]
> IThe "problem" or let's say difference of Brouter is that the avoidance of primary, secondary or teritiary roads is always interpreted as almost strict avoidance. 

Yes, but that issue can be fixed in the profile, so it's not strictly a problem with BRouter.

> They are not strict as in default BRouter profiles.

Setting the maxspeed on these roads to 1 km/h is pretty strict and indeed problematic. I would highly recommend to set those limits as high as possible while still achieving the effect you're looking for.

**Reply by devemux86** [reply]
Testing this [route](https://brouter.de/brouter-web/#map=10/48.8155/7.7948/standard&lonlats=8.078114,48.87915;7.218002,48.810498&profile=car-fast) with avoid motorways, BRouter car-fast has:
```
min if is_avoided_motorway then 0 else 999
```
We can make it not completely strict with:
```
min if is_avoided_motorway then 1 else 999
```
If we start making it large, the avoid uses motorways:
```
min if is_avoided_motorway then 60 else 999
```

@quaelnix is there a linear relationship with this value and avoid strength?
Like a percentage of a maximum value?

**Reply by devemux86** [reply]
> Setting the maxspeed on these roads to 1 km/h is pretty strict and indeed problematic. I would highly recommend to set those limits as high as possible while still achieving the effect you're looking for.

@quaelnix what value do you mean?

(please also see my above comment)

**Comment by quaelnix** [comment]
> is there a linear relationship with this value and avoid strength?

Yes, but the avoidance strength is determined by the ratio between the `maxspeed` of that road segment and the `maxspeed` of all other road segments that are considered for the final route. That's also the reason why you cannot simply copy and paste weights between profiles with a different cost structure.

> what value do you mean?

`min if is_avoided_motorway then 1 else 999` the unit of these two numbers is `km/h`. So if you for example set the maxspeed of a 1 km long segment of road to 1 km/h then BRouter would happily do a detour of up to 70 km length if there is an alternative road where you can drive 70 km/h.

**Comment by DolomitesTom** [comment]
Have tested again with 

**consider town**
avoid motorways
avoid urban
valid for motorcycle
avoid upaved
vmax = 160

Still the boring west route is the result. When I uncheck "consider town" and leave the "forest" only, then it really drives through the forest, but also takes the D1420 main road. No problem, one normally sets some shaping points anyway. But with unchecking consider_town it again takes the D459 instead of D1059

![grafik](https://github.com/devemux86/cruiser/assets/121371629/9ebd93cb-9da0-49f8-9f2d-4d22ed97d4dc)

When I active consider_town together with consider_forest, then it takes the D1059, but then again it would take boring west route (in this case).
![grafik](https://github.com/devemux86/cruiser/assets/121371629/bda2f1a4-ad8a-4ffe-9415-d8db0e127709)

It seems that these new options can't be used in combination. But normally I also want to avoid cities on a motorcycle.

With all new options active (I would like to use this on a motorcycle if it works), I get a pretty good route
![grafik](https://github.com/devemux86/cruiser/assets/121371629/7391f752-a18a-4abe-9249-eb97267ec351)

The strange thing is: if you only then additionally activate avoid_toll, it makes again a weird route what is not pursuable because there is no toll road there far and wide in the Vosges.
![grafik](https://github.com/devemux86/cruiser/assets/121371629/e7c33c91-aba5-4861-b90a-c39368ccac3a)
In this situation I would expect that avoid_toll has no influence at all on this route and should not lead to a change.

Can it be that we need a kind of weighting for these new options, as well as something like mitigated levels of avoidance from 0 to 6 e.g. from no avoidance over weak avoidance till strong avoidance for each of these main road types (primary, secondary, tertiary). I guess this all in combination would get the perfect route, just loud wishful thinking, don't know if technically possible. In any case, there is clearly more variety in it.

**Reply by devemux86** [reply]
> Can it be that we need a kind of weighting for these new options, as well as something like mitigated levels of avoidance from 0 to 6 e.g. from no avoidance over weak avoidance till strong avoidance for each of these main road types (primary, secondary, tertiary).

Looks like it's not so straightforward, see @quaelnix's answer above: https://github.com/devemux86/cruiser/discussions/266#discussioncomment-6646461
However, even so we must examine the possibility of avoidance strenghts.

**Comment by quaelnix** [comment]
@DolomitesTom, I absolutely agree with your problem analysis. @devemux86 as an inspiration:
```
assign no_river_penalty
  switch consider_river
    switch not estimated_river_class=4|5|6 99 0 0

assign no_forest_penalty
  switch consider_forest
    switch not estimated_forest_class=4|5|6 99 0 0

assign town_penalty
  switch consider_town
    switch not estimated_town_class=4|5|6 99 0 0

assign maxspeed =
  min if greater town_penalty 0 then 30 else 999
  min if greater no_forest_penalty 0 then 10 else 999
  min if greater no_river_penalty 0 then 10 else 999
  min if greater noise_penalty 0 then 15 else 999
  min if greater traffic_penalty 0 then 30 else 999
  min if is_avoided_toll_road then 1 else 999
  min if is_avoided_motorway then 1 else 999
  min if is_avoided_ferry then 1 else 999
  min if is_avoided_primary then 1 else 999
  min if is_avoided_secondary then 1 else 999
  min if is_avoided_tertiary then 1 else 999
  min if is_avoided_urban then 10 else 999
  min onewayspeedlimit
  min accessspeedlimit
  min maxspeed_explicit
  min maxspeed_implicit
  min maxspeed_surface maxspeed_tracktype
```
![profile_tweaks](https://github.com/devemux86/cruiser/assets/122357328/89dbde78-585e-4e76-bb3b-6bd78b8e9c87)
Test route: [https://brouter.de/brouter-web/#map=10/48.5052/7.3742/standard&lonlats=...](https://brouter.de/brouter-web/#map=10/48.5052/7.3742/standard&lonlats=7.337113,48.752424;7.301687,48.256468)

---

But in order to really get this right you would probably have to rewrite this entire section.
```
# add estimate tags
assign noforestpenaltylimit
   switch consider_forest
     switch estimated_forest_class=   15
     switch estimated_forest_class=1  20
     switch estimated_forest_class=2  25
     switch estimated_forest_class=3  30
     switch estimated_forest_class=4  50
     switch estimated_forest_class=5  999
     switch estimated_forest_class=6  999 999 999

assign maxspeed =
  min noforestpenaltylimit
  min onewayspeedlimit
  min accessspeedlimit
  min maxspeed_explicit
  min maxspeed_implicit
  min maxspeed_surface maxspeed_tracktype
```

The lower the forest class, the lower the assumed `maxspeed`, the higher the indirect penalty.

**Reply by devemux86** [reply]
@quaelnix thanks for the help!

Can you explain a bit about e.g. the town change?
This doesn't seem to avoid all town classes (which we want)?

Are the changes in the other classes similar (avoid half the classes)?
Or I misunderstand the changes? :slightly_smiling_face: 

```
switch not estimated_town_class=4|5|6 99 0 0
```

**Reply by quaelnix** [reply]
> This doesn't seem to avoid all town classes (which we want)?

`switch not estimated_town_class=4|5|6 99 0 0` was a typo, because I copy and pasted the forest thing without thinking. :blush:

I would try to solve the town penalty like this:
```
# add estimate tags
assign townpenaltylimit
   switch consider_town
     switch estimated_town_class=   999
     switch estimated_town_class=1  40
     switch estimated_town_class=2  35
     switch estimated_town_class=3  30
     switch estimated_town_class=4  25
     switch estimated_town_class=5  20
     switch estimated_town_class=6  15 15 999

assign maxspeed =
  min townpenaltylimit
  min onewayspeedlimit
  min accessspeedlimit
  min maxspeed_explicit
  min maxspeed_implicit
  min maxspeed_surface maxspeed_tracktype
```
But that won't do anything in Alsace until we fix the town tag, because there are basically no cities with more than 50k inhabitants. With the current town logic you might as well just do it like this:
```
assign townpenaltylimit switch estimated_town_class= 999 30
```

> Are the changes in the other classes similar (avoid half the classes)?

The "avoid half the classes" hack is only needed for the forest tag and the reason it is needed is because unfortunately we currently treat a lot of things as forest that are not forest, see: [https://github.com/abrensch/...brouter.sql#L548-L549](https://github.com/abrensch/brouter/blob/e66468b/misc/scripts/mapcreation/brouter.sql#L548-L549). The hack is not perfect, but it at least makes the forest tag somewhat usable.

Edit: The "avoid half the classed" hack is also needed for the river tag, because right now even little ponds somewhere alongside the road are taken into account during the tag calculation, which really needs to change in the near future.

So basically:
- `estimated_river_class` ignore classes 1, 2 and 3
- `estimated_noise_class` works as one would except
- `estimated_town_class` more or less completely broken (which is really sad)
- `estimated_traffic_class` more or less working, but still very far from perfect
- `estimated_forest_class` ignore classes 1, 2 and 3 (and keep in mind that it is not just forests)

**Reply by devemux86** [reply]
@quaelnix 

> But in order to really get this right you would probably have to rewrite this entire section.
The lower the forest class, the lower the assumed maxspeed, the higher the indirect penalty.

I tested your suggestion, but then your example https://github.com/devemux86/cruiser/discussions/266#discussioncomment-6649053 does not go though the forest?
Compared to this rule that works better in this example:
```
switch not estimated_forest_class=4|5|6 99 0 0
```

**Reply by quaelnix** [reply]
I did not manually optimize the weights in [the example I gave above](https://github.com/devemux86/cruiser/discussions/266#discussioncomment-6649053). Try these:
```
     switch estimated_forest_class=   10
     switch estimated_forest_class=1  15
     switch estimated_forest_class=2  20
     switch estimated_forest_class=3  25
     switch estimated_forest_class=4  30
```
You may have to further decrease them, but I would always take small steps and test thoroughly.

> Compared to this rule that works better in this example:

```
switch not estimated_forest_class=4|5|6 99 0 0
```
this is a binary switch answering the question, are we in a forest or are we not.

The actual penalty was applied here:
```
min if greater no_forest_penalty 0 then 10 else 999
```
Where `maxspeed` was set to 10 km/h in case the road is inside of a forest (or something like a forest). In my example the "max penalty" was 15 km/h, which explains why it was less strong.

Edit: Link to the complete profile text: https://paste.rs/wJ72o

**Reply by devemux86** [reply]
@quaelnix thanks for the help!

I updated the driving-cruiser profile, if anyone wants to test:
- https://github.com/devemux86/cruiser/discussions/272

**Comment by quaelnix** [comment]
> In this situation I would expect that avoid_toll has no influence at all

Adding a penalty to all roads will change the ratio between the roads which will lead to different routes. That is also the reason why I chose a multiplicative approach in my own [gravel profile](https://github.com/quaelnix/brouter-profiles/blob/main/gravel.brf).

> I guess this all in combination would get the perfect route, just loud wishful thinking, don't know if technically possible.

It should be possible to get very close to perfection, but it will be a lot of work - and I personally would wait until the dust settles regarding the new pseudo-tags.

**Comment by devemux86** [comment]
Based on @quaelnix's nice recommendations I updated the driving-cruiser profile:
(you can also test it on BRouter website)
- https://github.com/devemux86/cruiser/discussions/272

The magic happens after line 222.

**Comment by Leeb64** [comment]
test this:
[driving-cruiser.txt.txt](https://github.com/devemux86/cruiser/files/12288839/driving-cruiser.txt.txt)
line 266(  min if is_avoided_urban then 5 else 999)
->original:(  min if is_avoided_urban then 10 else 999)

**Comment by devemux86** [comment]
Improved `driving-cruiser` profile based on @quaelnix suggestions is included in [Cruiser 3.0.29](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

It is also available for testing on [BRouter website](https://brouter.de/brouter-web/):
- https://github.com/devemux86/cruiser/discussions/272

**Reply by DolomitesTom** [reply]
With only a few shaping points it' generates already a pretty good route.

**Comment by Leeb64** [comment]
<img src="https://github.com/devemux86/cruiser/assets/120722053/d50cb99c-7021-4c32-ae93-298e399b6cd9" width="50%"/>

Without intermediate points.
Nearly perfect....
![testx](https://github.com/devemux86/cruiser/assets/120722053/ae38e62c-2b8a-4b79-9881-f3a34d581b70)
[driving-cruiser.txt.txt](https://github.com/devemux86/cruiser/files/12292621/driving-cruiser.txt.txt)

**Reply by devemux86** [reply]
@Leeb64 did you find that a more strict avoidance (5 instead of 10) for avoid_urban is needed?

**Reply by Leeb64** [reply]
5 is definitely better than 10 !

**Reply by DolomitesTom** [reply]
Which map design is it?

**Reply by Leeb64** [reply]
[__Normalstyle.zip](https://github.com/devemux86/cruiser/files/12292886/__Normalstyle.zip)
->Elevatex

**Reply by Leeb64** [reply]
the calculated travel time is probably quite wrong....
Half would be fine...

**Reply by devemux86** [reply]
@Leeb64 what do you have selected in "Settings | Routes | Options | Route duration"?

**Reply by devemux86** [reply]
@Leeb64 do you see the same route duration on BRouter website?

**Reply by Leeb64** [reply]
yes

**Reply by DolomitesTom** [reply]
Thx, is this elevatex something that devemux could offer for download?

**Reply by Leeb64** [reply]
It is a color "optimized" Elevate/Elements profile

**Comment by Leeb64** [comment]
Strange things happen with "Prefer forests and green regions". It routes (original profile) in towns through some small streets.
I changed that a bit...
![Screenshot 2023-08-09 at 08-47-40 BRouter Web Client Einstellungen](https://github.com/devemux86/cruiser/assets/120722053/b56b043e-723b-44d9-a3b2-8952864d3c49)
![Screenshot 2023-08-09 at 08-46-33 BRouter Web Client original](https://github.com/devemux86/cruiser/assets/120722053/8428e15e-e1b7-4809-ba71-fbfad82da647)
![Screenshot 2023-08-09 at 08-42-03 BRouter Web Client forest geändert](https://github.com/devemux86/cruiser/assets/120722053/879db73a-a15f-43ec-8af7-5ee528d40fe2)

If you prefer rivers and seas and avoid noise, something similar, counterproductive happens.....
I think the avoidance stuff like: rivers + sea, forest, noise and traffic should "switch off" when driving through towns.
You might have to create a kind of no-go area around localities. Different rules then apply in these areas....

**Reply by Leeb64** [reply]
It's going in the right direction.....but it may take a while...

**Comment by Leeb64** [comment]
I've been doctoring around on the profile again.
It brings very useful results. The avoidances: River/Noise/Traffic are to be enjoyed with caution...
[driving-cruiserxx.txt.txt](https://github.com/devemux86/cruiser/files/12303279/driving-cruiserxx.txt.txt)



I changed:
Line 163:  50->8
Line 240:   5->10
Line 267:   5->1
Line268:   10->5

**Comment by quaelnix** [comment]
> Strange things happen with "Prefer forests and green regions". It routes (original profile) in towns through some small streets.

@Leeb64, it's a known problem, and I've been trying for months to convince @EssBee59 that this behavior is bad. The situation with the town and forest tags is already better than it used to be, but still far from acceptable.

**Comment by devemux86** [comment]
Interesting recommendation by @EssBee59 in https://github.com/abrensch/brouter/pull/579#issuecomment-1674315124:

_"As recommended, never select "consider_river" and"consider_forest" together."_

**Comment by Leeb64** [comment]
I actually just wanted to ask whether anything has happened here regarding the quality of the avoidance criteria?
As it was, it was unusable...

**Reply by devemux86** [reply]
What specific BRouter avoid options?

The main avoid options use the road types as they exist in OpenStreetMap database.
If there is something wrong, please post route examples so we can examine the road tags.
This can be done in OpenStreetMap website or better in BRouter website, see instructions:
- https://github.com/devemux86/cruiser/discussions/272

---

The "Other" options (cities, forests, etc) are under development by the BRouter developers:
- https://github.com/abrensch/brouter/issues/608
- https://github.com/abrensch/brouter/pull/614

Better work with the main avoid options and not change anything in the "Other" options.

**Reply by Leeb64** [reply]
I meant exactly these “other” avoidance options.

**Comment by DolomitesTom** [comment]
By the way what about the other very useful option of different sectional routing options, for example use motorways only for a certain section. Is this something that could work in the future? In the motorcycle profile you normally have deactivated motorways. But in special situations it is the better option.

**Reply by devemux86** [reply]
This is already planned for the future. :slightly_smiling_face: 
- https://github.com/devemux86/cruiser/discussions/142

**Reply by DolomitesTom** [reply]
perfect.

**Comment by EssBee59** [comment]
Hello!

I am not motor-biker but after reading above some requirements to the
cruiser profile I think that the expected profile is not typical for car / the
KinematicModel?

In other words, did you consider or test the use of the Brouter with the
standard model?

An advantage is, by defining each cost manually ( with direct/transparent
impact on the routing), a high flexibility is provided for the routing
("avoid" or "forbid" a way by setting high or very-high cost on tags of
that way)
I attach an example of such a profile in the case you are interested to
test.
(but it is not completed, many parameters are missing such as "access" for
motorbike, avoid_toll etc...)
[cruiser_classic.brf.txt](https://github.com/devemux86/cruiser/files/12920543/cruiser_classic.brf.txt)

**Reply by devemux86** [reply]
Hi @EssBee59!

Are you one of the current BRouter developers?

> In other words, did you consider or test the use of the Brouter with the
standard model?

In the past I had modified with extra options one of @poutnikl profiles: https://github.com/poutnikl/Brouter-profiles
Do you mean something like that?

Many thanks for your effort! I will check the profile.

May I ask if you used a default profile as base for the above one?

**Reply by EssBee59** [reply]
Yes, I developped in the last monthes the pseudo-tags (noise, river, forest, etc,,,)

yes, the above profile is based on my profiles racebike-verylowtraffic and trekking_SB that can be used in my test-instance
http://brouter.de/essbee
